src/algebra/group/unique_prods.lean

```diff
@@ -0,0 +1,223 @@
+  unique_mul (A.image f) (B.image f) (f a0) (f b0) ↔ unique_mul A B a0 b0 :=
+begin
+  refine ⟨λ h, _, λ h, _⟩,
+  { intros a b ha hb ab,
+    rw [← hf.eq_iff, ← hf.eq_iff],
+    rw [← hf.eq_iff, map_mul, map_mul] at ab,
+    exact h (finset.mem_image.mpr ⟨_, ha, rfl⟩) (finset.mem_image.mpr ⟨_, hb, rfl⟩) ab},
+  { intros a b aA bB ab,
+    obtain ⟨a, ha, rfl⟩ : ∃ a' ∈ A, f a' = a := finset.mem_image.mp aA,
+    obtain ⟨b, hb, rfl⟩ : ∃ b' ∈ B, f b' = b := finset.mem_image.mp bB,
+    rw [hf.eq_iff, hf.eq_iff],
+    rw [← map_mul, ← map_mul, hf.eq_iff] at ab,
+    exact h ha hb ab },
+end
+
+/--  `unique_mul` is preserved under embeddings that are multiplicative.
+
+See `unique_mul.mul_hom_image_iff` for a version with swapped bundling. -/
+@[to_additive "`unique_add` is preserved under embeddings that are additive.
+
+See `unique_add.add_hom_image_iff` for a version with swapped bundling."]
+lemma mul_hom_map_iff (f : G ↪ H) (mul : ∀ x y, f (x * y) = f x * f y) :
+  unique_mul (A.map f) (B.map f) (f a0) (f b0) ↔ unique_mul A B a0 b0 :=
+begin
+  classical,
+  convert mul_hom_image_iff ⟨f, mul⟩ f.2;
+  { ext,
+    simp only [finset.mem_map, mul_hom.coe_mk, finset.mem_image] },
+end
+
+end unique_mul
+
+/--  Let `G` be a Type with addition.  `unique_sums G` asserts that any two non-empty
+finite subsets of `A` have the `unique_add` property, with respect to some element of their
+sum `A + B`. -/
+class unique_sums (G) [has_add G] : Prop :=
+( adds : ∀ {A B : finset G} (hA : A.nonempty) (hB : B.nonempty),
+  ∃ (a0 ∈ A) (b0 ∈ B), unique_add A B a0 b0)
+
+/--  Let `G` be a Type with multiplication.  `unique_prods G` asserts that any two non-empty
+finite subsets of `G` have the `unique_mul` property, with respect to some element of their
+product `A * B`. -/
+class unique_prods (G) [has_mul G] : Prop :=
+( muls : ∀ {A B : finset G} (hA : A.nonempty) (hB : B.nonempty),
+  ∃ (a0 ∈ A) (b0 ∈ B), unique_mul A B a0 b0 )
+
+attribute [to_additive] unique_prods
+
+namespace multiplicative
+
+instance {M} [has_add M] [unique_sums M] : unique_prods (multiplicative M) :=
+{ muls := λ A B hA hB, let A' : finset M := A in have hA': A'.nonempty := hA, by
+    { obtain ⟨a0, hA0, b0, hB0, J⟩ := unique_sums.adds hA' hB,
+      exact ⟨of_add a0, hA0, of_add b0, hB0, λ a b aA bB H, J aA bB H⟩ } }
+
+end multiplicative
+
+namespace additive
+
+instance {M} [has_mul M] [unique_prods M] : unique_sums (additive M) :=
+{ adds := λ A B hA hB, let A' : finset M := A in have hA': A'.nonempty := hA, by
+    { obtain ⟨a0, hA0, b0, hB0, J⟩ := unique_prods.muls hA' hB,
+      exact ⟨of_mul a0, hA0, of_mul b0, hB0, λ a b aA bB H, J aA bB H⟩ } }
+
+end additive
+
+--  please, golf me and rename me!
+@[to_additive] lemma eq_and_eq_of_le_of_le_of_mul_eq {A} [has_mul A] [linear_order A]
+  [covariant_class A A (*) (≤)] [covariant_class A A (function.swap (*)) (<)]
+  [contravariant_class A A (*) (≤)]
+  {a b a0 b0 : A} (ha : a0 ≤ a) (hb : b0 ≤ b) (ab : a * b = a0 * b0) :
```

Comment:
Does this help to golf the proof?
```suggestion
  {a b a0 b0 : A} (ha : a0 ≤ a) (hb : b0 ≤ b) (ab : a * b ≤ a0 * b0) :
```

Reply:
Anne, thanks!  I implemented your suggestion, although I am a little squeezed for time at the moment.  I will see how it works, but may not be too quick to react.

Reply:
@Vierkantor the build succeeded!